packages/babel-compiler/babel-compiler.js

```diff
@@ -216,40 +222,58 @@ BCp._inferHelper = function (inputFile, babelOptions, babelrc) {
     function req(id) {
       var isTopLevel = "./".indexOf(id.charAt(0)) < 0;
       var presetOrPlugin;
+      var presetOrPluginMeta;
 
       if (isTopLevel) {
         try {
           // If the identifier is top-level, try to prefix it with
           // "babel-plugin-" or "babel-preset-".
           presetOrPlugin = inputFile.require(prefix + id);
+          presetOrPluginMeta = inputFile.require(
+            packageNameFromTopLevelModuleId(prefix + id) + '/package.json');
         } catch (e) {
           if (e.code !== "MODULE_NOT_FOUND") {
             throw e;
           }
           // Fall back to requiring the plugin as-is if the prefix failed.
           presetOrPlugin = inputFile.require(id);
+          presetOrPluginMeta = inputFile.require(
+            packageNameFromTopLevelModuleId(id) + '/package.json');
         }
       } else {
         // If the identifier is not top-level, but relative or absolute,
         // then it will be required as-is, so that you can implement your
         // own Babel plugins locally, rather than always using plugins
         // installed from npm.
         presetOrPlugin = inputFile.require(id);
```

Comment:
Yeah `func.toString()` wouldn't have been my first choice but I wanted to avoid reading the file a 2nd time, maybe I'm a little conservative like that.  Without using `Module._{cache,compile}` directly.  But agree reading the file directly is cleaner and hopefully the OS will have it cached too.  Ultimately though this will depend on our discussion about how to best handle caching for this case.